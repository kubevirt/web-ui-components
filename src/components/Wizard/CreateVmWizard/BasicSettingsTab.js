import React from 'react';
import PropTypes from 'prop-types';
import { get } from 'lodash';

import {
  FormFactory,
  validateForm,
  getFieldValidation,
  DROPDOWN,
  CHECKBOX,
  TEXT_AREA,
  POSITIVE_NUMBER,
} from '../../Form';
import { getName, getMemory, getCpu, getCloudInitUserData } from '../../../utils/selectors';
import { getTemplate, getTemplateProvisionSource } from '../../../utils/templates';
import { validateDNS1123SubdomainValue, validateURL, validateContainer } from '../../../utils/validations';
import {
  NO_TEMPLATE,
  HELP_PROVISION_SOURCE_URL,
  HELP_PROVISION_SOURCE_PXE,
  HELP_PROVISION_SOURCE_CONTAINER,
  HELP_PROVISION_SOURCE_IMPORT,
  HELP_OS,
  HELP_FLAVOR,
  HELP_MEMORY,
  HELP_CPU,
  HELP_WORKLOAD,
} from './strings';

import {
  getFlavors,
  getOperatingSystems,
  getWorkloadProfiles,
  isFlavorType,
  isImageSourceType,
  selectVm,
  getTemplateFlavors,
  getTemplateWorkloadProfiles,
  getTemplateOperatingSystems,
  settingsValue,
  getV2VVmwareName,
} from '../../../k8s/selectors';

import {
  CUSTOM_FLAVOR,
  PROVISION_SOURCE_PXE,
  PROVISION_SOURCE_CONTAINER,
  PROVISION_SOURCE_URL,
  PROVISION_SOURCE_IMPORT,
  TEMPLATE_TYPE_VM,
} from '../../../constants';

import {
  NAME_KEY,
  NAMESPACE_KEY,
  DESCRIPTION_KEY,
  PROVISION_SOURCE_TYPE_KEY,
  CONTAINER_IMAGE_KEY,
  IMAGE_URL_KEY,
  USER_TEMPLATE_KEY,
  OPERATING_SYSTEM_KEY,
  FLAVOR_KEY,
  MEMORY_KEY,
  CPU_KEY,
  WORKLOAD_PROFILE_KEY,
  START_VM_KEY,
  CLOUD_INIT_KEY,
  USE_CLOUD_INIT_CUSTOM_SCRIPT_KEY,
  CLOUD_INIT_CUSTOM_SCRIPT_KEY,
  HOST_NAME_KEY,
  AUTHKEYS_KEY,
  BATCH_CHANGES_KEY,
} from './constants';

import { importProviders } from './providers';
import { V2VVMwareModel } from '../../../models';

const getProvisionSourceHelp = basicSettings => {
  const provisionSource = settingsValue(basicSettings, PROVISION_SOURCE_TYPE_KEY);
  switch (provisionSource) {
    case PROVISION_SOURCE_URL:
      return HELP_PROVISION_SOURCE_URL;
    case PROVISION_SOURCE_PXE:
      return HELP_PROVISION_SOURCE_PXE;
    case PROVISION_SOURCE_CONTAINER:
      return HELP_PROVISION_SOURCE_CONTAINER;
    case PROVISION_SOURCE_IMPORT:
      return HELP_PROVISION_SOURCE_IMPORT;
    default:
      return null;
  }
};

export const getFormFields = (
  basicSettings,
  namespaces,
  templates,
  selectedNamespace,
  createTemplate,
  WithResources,
  k8sCreate,
  k8sGet,
  k8sPatch
) => {
  const userTemplate = get(basicSettings[USER_TEMPLATE_KEY], 'value');
  const workloadProfiles = getWorkloadProfiles(basicSettings, templates, userTemplate);
  const operatingSystems = getOperatingSystems(basicSettings, templates, userTemplate);
  const flavors = getFlavors(basicSettings, templates, userTemplate);
  const imageSources = [PROVISION_SOURCE_PXE, PROVISION_SOURCE_URL, PROVISION_SOURCE_CONTAINER];
  const userTemplateNames = getTemplate(templates, TEMPLATE_TYPE_VM).map(getName);
  userTemplateNames.push(NO_TEMPLATE);

  let namespaceDropdown;
  let startVmCheckbox;
  let userTemplateDropdown;
  let providersSection = {};

  if (!selectedNamespace) {
    namespaceDropdown = {
      id: 'namespace-dropdown',
      title: 'Namespace',
      type: DROPDOWN,
      defaultValue: '--- Select Namespace ---',
      choices: namespaces.map(getName),
      required: true,
    };
  }

  if (!createTemplate) {
    startVmCheckbox = {
      id: 'start-vm',
      title: 'Start virtual machine on creation',
      type: CHECKBOX,
      noBottom: true,
    };
    userTemplateDropdown = {
      id: 'template-dropdown',
      title: 'Template',
      type: DROPDOWN,
      defaultValue: '--- Select Template ---',
      choices: userTemplateNames,
    };

    providersSection = importProviders(basicSettings, WithResources, k8sCreate, k8sGet, k8sPatch);
    imageSources.push(PROVISION_SOURCE_IMPORT);
  }

  return {
    [NAME_KEY]: {
      id: 'vm-name',
      title: 'Name',
      required: true,
      validate: settings => validateDNS1123SubdomainValue(settingsValue(settings, NAME_KEY)),
    },
    [DESCRIPTION_KEY]: {
      id: 'vm-description',
      title: 'Description',
      type: TEXT_AREA,
    },
    [NAMESPACE_KEY]: namespaceDropdown,
    [USER_TEMPLATE_KEY]: userTemplateDropdown,
    [PROVISION_SOURCE_TYPE_KEY]: {
      id: 'image-source-type-dropdown',
      title: 'Provision Source',
      type: DROPDOWN,
      defaultValue: '--- Select Provision Source ---',
      choices: imageSources,
      required: true,
      disabled: userTemplate !== undefined,
      help: getProvisionSourceHelp(basicSettings),
    },
    [CONTAINER_IMAGE_KEY]: {
      id: 'provision-source-container',
      title: 'Container Image',
      required: true,
      isVisible: basicVmSettings => isImageSourceType(basicVmSettings, PROVISION_SOURCE_CONTAINER),
      disabled: userTemplate !== undefined,
      validate: settings => validateContainer(settingsValue(settings, CONTAINER_IMAGE_KEY)),
    },
    [IMAGE_URL_KEY]: {
      id: 'provision-source-url',
      title: 'URL',
      required: true,
      isVisible: basicVmSettings => isImageSourceType(basicVmSettings, PROVISION_SOURCE_URL),
      disabled: userTemplate !== undefined,
      validate: settings => validateURL(settingsValue(settings, IMAGE_URL_KEY)),
    },
    ...providersSection,
    [OPERATING_SYSTEM_KEY]: {
      id: 'operating-system-dropdown',
      title: 'Operating System',
      type: DROPDOWN,
      defaultValue: '--- Select Operating System ---',
      choices: operatingSystems,
      required: true,
      disabled: userTemplate !== undefined,
      help: HELP_OS,
    },
    [FLAVOR_KEY]: {
      id: 'flavor-dropdown',
      title: 'Flavor',
      type: DROPDOWN,
      defaultValue: '--- Select Flavor ---',
      choices: flavors,
      required: true,
      disabled: userTemplate !== undefined && flavors.length === 1,
      help: HELP_FLAVOR,
    },
    [MEMORY_KEY]: {
      id: 'resources-memory',
      title: 'Memory (GB)',
      type: POSITIVE_NUMBER,
      required: true,
      isVisible: basicVmSettings => isFlavorType(basicVmSettings, CUSTOM_FLAVOR),
      help: HELP_MEMORY,
    },
    [CPU_KEY]: {
      id: 'resources-cpu',
      title: 'CPUs',
      type: POSITIVE_NUMBER,
      required: true,
      isVisible: basicVmSettings => isFlavorType(basicVmSettings, CUSTOM_FLAVOR),
      help: HELP_CPU,
    },
    [WORKLOAD_PROFILE_KEY]: {
      id: 'workload-profile-dropdown',
      title: 'Workload Profile',
      type: DROPDOWN,
      defaultValue: '--- Select Workload Profile ---',
      choices: workloadProfiles,
      required: true,
      help: HELP_WORKLOAD,
      disabled: userTemplate !== undefined,
    },
    [START_VM_KEY]: startVmCheckbox,
    [CLOUD_INIT_KEY]: {
      id: 'use-cloud-init',
      title: 'Use cloud-init',
      type: CHECKBOX,
    },
    [USE_CLOUD_INIT_CUSTOM_SCRIPT_KEY]: {
      id: 'use-cloud-init-custom-script',
      title: 'Use custom script',
      type: CHECKBOX,
      isVisible: basicVmSettings => settingsValue(basicVmSettings, CLOUD_INIT_KEY, false),
    },
    [HOST_NAME_KEY]: {
      id: 'cloud-init-hostname',
      title: 'Hostname',
      isVisible: basicVmSettings =>
        settingsValue(basicVmSettings, CLOUD_INIT_KEY, false) &&
        !settingsValue(basicVmSettings, USE_CLOUD_INIT_CUSTOM_SCRIPT_KEY, false),
    },
    [AUTHKEYS_KEY]: {
      id: 'cloud-init-ssh',
      title: 'Authenticated SSH Keys',
      type: TEXT_AREA,
      isVisible: basicVmSettings =>
        settingsValue(basicVmSettings, CLOUD_INIT_KEY, false) &&
        !settingsValue(basicVmSettings, USE_CLOUD_INIT_CUSTOM_SCRIPT_KEY, false),
    },
    [CLOUD_INIT_CUSTOM_SCRIPT_KEY]: {
      id: 'cloud-init-custom-script',
      title: 'Custom Script',
      type: TEXT_AREA,
      className: 'kubevirt-create-vm-wizard__custom-cloud-script-textarea',
      isVisible: basicVmSettings =>
        settingsValue(basicVmSettings, CLOUD_INIT_KEY, false) &&
        settingsValue(basicVmSettings, USE_CLOUD_INIT_CUSTOM_SCRIPT_KEY, false),
    },
  };
};

const asValueObject = (value, validation) => ({
  value,
  validation,
});

// TODO: To avoid race conditions as the basicSettings tab is bound to the CreateVmWizard state at the render-time, it would
//  be better to retrieve basicSettings at the time of its actual use - to align behavior with the setState().
// The onChange() bellow should be called with just the diff, not whole copy of the stepData.
const publish = ({ basicSettings, templates, onChange, dataVolumes }, value, target, formValid, formFields) => {
  let newBasicSettings;
  if (target === BATCH_CHANGES_KEY) {
    // the "value" is an array of pairs { value, target }
    const difference = value.value.reduce((map, obj) => {
      map[obj.target] = { value: obj.value }; // validation will be set in a next step
      return map;
    }, {});
    newBasicSettings = {
      ...basicSettings,
      ...difference,
    };
    value.value.forEach(pair => {
      // set field validation
      newBasicSettings[pair.target].validation = getFieldValidation(
        formFields[pair.target],
        pair.value,
        newBasicSettings
      );
    });
    formValid = validateForm(formFields, newBasicSettings);
  } else {
    newBasicSettings = {
      ...basicSettings,
      [target]: value,
    };

    if (target === USER_TEMPLATE_KEY) {
      if (value.value === NO_TEMPLATE) {
        newBasicSettings[target] = asValueObject(undefined);
      } else {
        const allTemplates = getTemplate(templates, TEMPLATE_TYPE_VM);
        if (allTemplates.length > 0) {
          const userTemplate = allTemplates.find(template => template.metadata.name === value.value);
          updateTemplateData(userTemplate, newBasicSettings, dataVolumes);
        }
      }
      formValid = validateForm(formFields, newBasicSettings);
    }
  }

  onChange(newBasicSettings, formValid);
};

const updateTemplateData = (userTemplate, newBasicSettings, dataVolumes) => {
  if (userTemplate) {
    const vm = selectVm(userTemplate.objects);

    // update flavor
    const [flavor] = getTemplateFlavors([userTemplate]);
    newBasicSettings[FLAVOR_KEY] = asValueObject(flavor);
    if (flavor === CUSTOM_FLAVOR) {
      newBasicSettings.cpu = asValueObject(getCpu(vm));
      const memory = getMemory(vm);
      newBasicSettings.memory = memory ? asValueObject(parseInt(memory, 10)) : undefined;
    }

    // update os
    const [os] = getTemplateOperatingSystems([userTemplate]);
    newBasicSettings[OPERATING_SYSTEM_KEY] = asValueObject(os);

    // update workload profile
    const [workload] = getTemplateWorkloadProfiles([userTemplate]);
    newBasicSettings[WORKLOAD_PROFILE_KEY] = asValueObject(workload);

    // update cloud-init
    const cloudInitUserData = getCloudInitUserData(vm);
    if (cloudInitUserData) {
      newBasicSettings[CLOUD_INIT_KEY] = asValueObject(true);
      newBasicSettings[USE_CLOUD_INIT_CUSTOM_SCRIPT_KEY] = asValueObject(true);
      newBasicSettings[CLOUD_INIT_CUSTOM_SCRIPT_KEY] = asValueObject(cloudInitUserData || '');
    } else if (get(newBasicSettings[CLOUD_INIT_KEY], 'value')) {
      newBasicSettings[CLOUD_INIT_KEY] = asValueObject(false);
      newBasicSettings[USE_CLOUD_INIT_CUSTOM_SCRIPT_KEY] = asValueObject(false);
    }

    // update provision source
    const provisionSource = getTemplateProvisionSource(userTemplate, dataVolumes);
    if (provisionSource) {
      newBasicSettings[PROVISION_SOURCE_TYPE_KEY] = asValueObject(provisionSource.type);
      switch (provisionSource.type) {
        case PROVISION_SOURCE_CONTAINER:
          newBasicSettings[CONTAINER_IMAGE_KEY] = asValueObject(provisionSource.source);
          break;
        case PROVISION_SOURCE_URL:
          newBasicSettings[IMAGE_URL_KEY] = asValueObject(provisionSource.source);
          break;
        case PROVISION_SOURCE_PXE:
          break;
        default:
          // eslint-disable-next-line
          console.warn(`Unknown provision source ${provisionSource.type}`);
          break;
      }
    } else {
      // eslint-disable-next-line
      console.warn(`Cannot detect provision source for template ${getName(userTemplate)}`);
    }
  }
  return newBasicSettings;
};

// Do clean-up
export const onCloseBasic = async (basicSettings, callerContext) => {
  const v2vvmwareName = getV2VVmwareName(basicSettings);
  if (v2vvmwareName) {
    // This is a friendly help to keep things clean.
    // If missed here (e.g. when the browser window is closed), the kubevirt-vmware controller's garbage
    // collector will do following automatically after a delay.
    const resource = {
      metadata: {
        name: v2vvmwareName,
        // TODO: potential issue if the user changed the namespace after creation of the v2vvmware object
        // to fix: either store the namespace along the v2vvmwareName or empty v2vvmwareName on namespace change
        namespace: settingsValue(basicSettings, NAMESPACE_KEY),
      },
    };
    try {
      await callerContext.k8sKill(V2VVMwareModel, resource);
    } catch (error) {
      // eslint-disable-next-line no-console
      console.log(
        'Failed to remove temporary V2VVMWare object. It is not an issue, it will be garbage collected later if still present, resource: ',
        resource,
        ', error: ',
        error
      );
    }
  }
};

export class BasicSettingsTab extends React.Component {
  constructor(props) {
    super(props);
    if (props.selectedNamespace) {
      this.updateSelectedNamespace(props);
    }
  }

  componentDidUpdate(prevProps) {
    const newNamespace = this.props.selectedNamespace;
    const oldNamespace = prevProps.selectedNamespace;

    if (getName(newNamespace) !== getName(oldNamespace)) {
      this.updateSelectedNamespace(this.props);
    }
  }

  updateSelectedNamespace = props => {
    const {
      basicSettings,
      namespaces,
      selectedNamespace,
      templates,
      createTemplate,
      WithResources,
      k8sCreate,
      k8sGet,
      k8sPatch,
    } = props;
    const formFields = getFormFields(
      basicSettings,
      namespaces,
      templates,
      selectedNamespace,
      createTemplate,
      WithResources,
      k8sCreate,
      k8sGet,
      k8sPatch
    );
    const valid = validateForm(formFields, basicSettings);
    publish(props, asValueObject(getName(selectedNamespace)), NAMESPACE_KEY, valid, formFields);
  };

  render() {
    const {
      basicSettings,
      namespaces,
      templates,
      selectedNamespace,
      createTemplate,
      WithResources,
      k8sCreate,
      k8sGet,
      k8sPatch,
    } = this.props;
    const formFields = getFormFields(
      basicSettings,
      namespaces,
      templates,
      selectedNamespace,
      createTemplate,
      WithResources,
      k8sCreate,
      k8sGet,
      k8sPatch
    );

    return (
      <FormFactory
        fields={formFields}
        fieldsValues={basicSettings}
        onFormChange={(newValue, target, formValid) => publish(this.props, newValue, target, formValid, formFields)}
      />
    );
  }
}

BasicSettingsTab.defaultProps = {
  selectedNamespace: undefined,
  createTemplate: false,
};

BasicSettingsTab.propTypes = {
  WithResources: PropTypes.func.isRequired,
  k8sCreate: PropTypes.func.isRequired,
  k8sGet: PropTypes.func.isRequired,
  k8sPatch: PropTypes.func.isRequired,
  templates: PropTypes.array.isRequired,
  namespaces: PropTypes.array.isRequired,
  selectedNamespace: PropTypes.object, // used only in initialization
  basicSettings: PropTypes.object.isRequired,
  // eslint-disable-next-line react/no-unused-prop-types
  onChange: PropTypes.func.isRequired,
  createTemplate: PropTypes.bool,
  // eslint-disable-next-line react/no-unused-prop-types
  dataVolumes: PropTypes.array.isRequired,
};
