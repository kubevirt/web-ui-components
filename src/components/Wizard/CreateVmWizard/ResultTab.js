import React from 'react';
import PropTypes from 'prop-types';

import { Loading } from '../../Loading';

const getCreationText = (state, createTemplate) => `Creation of VM ${createTemplate ? 'Template ' : ''}${state}`;

export const ResultTab = ({ isSuccessful, isCreateTemplate, children }) => {
  if (isSuccessful == null) {
    return <Loading key="progress" text={getCreationText('in progress', isCreateTemplate)} />;
  }
  const state = isSuccessful ? 'was successful' : 'failed';
  const icon = isSuccessful ? 'pficon-ok' : 'pficon-error-circle-o';

  return (
    <div className="wizard-pf-complete blank-slate-pf" key="success">
      <div className="wizard-pf-success-icon">
        <span className={`pficon ${icon}`} />
      </div>
      <h3 className="blank-slate-pf-main-action">{getCreationText(state, isCreateTemplate)}</h3>
      {children}
    </div>
  );
};

ResultTab.defaultProps = {
  children: null,
  isSuccessful: null,
  isCreateTemplate: false,
};

ResultTab.propTypes = {
  children: PropTypes.node,
  isSuccessful: PropTypes.bool,
  isCreateTemplate: PropTypes.bool,
};
