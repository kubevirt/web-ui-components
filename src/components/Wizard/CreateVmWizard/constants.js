export const BASIC_SETTINGS_TAB_IDX = 0;
export const NETWORK_TAB_IDX = 1;
export const STORAGE_TAB_IDX = 2;
export const RESULTS_TAB_IDX = 3;

export const ALL_TABS = [BASIC_SETTINGS_TAB_IDX, NETWORK_TAB_IDX, STORAGE_TAB_IDX, RESULTS_TAB_IDX];

// BasicSettingsTab
export const NAME_KEY = 'name';
export const NAMESPACE_KEY = 'namespace';
export const DESCRIPTION_KEY = 'description';
export const IMAGE_SOURCE_TYPE_KEY = 'imageSourceType';
export const REGISTRY_IMAGE_KEY = 'registryImage';
export const IMAGE_URL_KEY = 'imageURL';
export const USER_TEMPLATE_KEY = 'userTemplate';
export const OPERATING_SYSTEM_KEY = 'operatingSystem';
export const FLAVOR_KEY = 'flavor';
export const MEMORY_KEY = 'memory';
export const CPU_KEY = 'cpu';
export const WORKLOAD_PROFILE_KEY = 'workloadProfile';
export const START_VM_KEY = 'startVM';
export const CREATE_TEMPLATE_KEY = 'createTemplate ';
export const CLOUD_INIT_KEY = 'cloudInit';
export const HOST_NAME_KEY = 'hostname';
export const AUTHKEYS_KEY = 'authKeys';
