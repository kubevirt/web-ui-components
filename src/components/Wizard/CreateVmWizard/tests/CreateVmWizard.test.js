import React from 'react';
import { shallow } from 'enzyme';
import { WizardPattern } from 'patternfly-react';
import { CreateVmWizard } from '../CreateVmWizard';
import { templates, networkConfigs } from '../../../../constants';
import {
  namespaces,
  persistentVolumeClaims,
  storageClasses,
  units
} from '../../NewVmWizard/fixtures/NewVmWizard.fixture';

import { validBasicSettings } from '../fixtures/BasicSettingsTab.fixture';
import { createVM } from '../../../../k8s/request';
import { BASIC_SETTINGS_TAB_IDX, NETWORK_TAB_IDX, STORAGE_TAB_IDX, RESULTS_TAB_IDX, ALL_TABS } from '../constants';

jest.mock('../../../../k8s/request');

const testCreateVmWizard = () => (
  <CreateVmWizard
    onHide={() => {}}
    templates={templates}
    namespaces={namespaces}
    k8sCreate={() => {}}
    persistentVolumeClaims={persistentVolumeClaims}
    storageClasses={storageClasses}
    units={units}
    networkConfigs={networkConfigs}
  />
);

const testWalkThrough = () => {
  const component = shallow(testCreateVmWizard());

  component.instance().onStepDataChanged(validBasicSettings, true);
  expect(component.find(WizardPattern).props().nextStepDisabled).toBeFalsy();

  component.instance().onStepChanged(NETWORK_TAB_IDX); // should allow going forward
  expect(component.state().activeStepIndex).toEqual(NETWORK_TAB_IDX);

  component.instance().onStepChanged(BASIC_SETTINGS_TAB_IDX); // try to go back
  expect(component.state().activeStepIndex).toEqual(BASIC_SETTINGS_TAB_IDX);

  component.instance().onStepChanged(NETWORK_TAB_IDX); // forward
  expect(component.find(WizardPattern).props().nextText).toBe('Next');
  component.instance().onStepChanged(STORAGE_TAB_IDX);

  expect(component.state().activeStepIndex).toEqual(STORAGE_TAB_IDX);
  expect(component.find(WizardPattern).props().nextText).toBe('Create Virtual Machine');
  expect(component.instance().lastStepReached()).toBeFalsy();

  component.instance().onStepDataChanged([{}], false); // create empty disk
  expect(component.find(WizardPattern).props().nextStepDisabled).toBeTruthy();
  component.instance().onStepDataChanged(
    [
      {
        id: 1,
        name: 'D',
        size: '15',
        storageClass: 'iscsi',
        isBootable: true
      },
      {
        id: 2,
        isBootable: false,
        attachStorage: {
          id: 'disk-two',
          name: 'disk Two',
          size: '15',
          storageClass: 'glusterfs'
        }
      }
    ],
    true
  );

  expect(createVM).not.toHaveBeenCalled();
  component.instance().onStepChanged(RESULTS_TAB_IDX); // create vm
  expect(component.state().activeStepIndex).toEqual(RESULTS_TAB_IDX);
  expect(component.instance().lastStepReached()).toBeTruthy();
  expect(createVM).toHaveBeenCalled();
  expect(component.find(WizardPattern).props().previousStepDisabled).toBeTruthy();
  component.instance().onStepChanged(BASIC_SETTINGS_TAB_IDX); // should not allow going backwards
  expect(component.state().activeStepIndex).toEqual(RESULTS_TAB_IDX);
};

describe('<CreateVmWizard />', () => {
  beforeEach(() => {
    createVM.mockClear();
  });

  it('renders correctly', () => {
    const component = shallow(testCreateVmWizard());
    expect(component).toMatchSnapshot();
  });

  it('is visible when mounted', () => {
    const component = shallow(testCreateVmWizard());
    expect(component.find(WizardPattern).props().show).toBeTruthy();
  });

  it("onStepChanged doesn't update activeStepIndex due to invalid form", () => {
    createVM.mockReturnValueOnce(new Promise((resolve, reject) => resolve({ result: 'VM created' })));
    const component = shallow(testCreateVmWizard());
    expect(component.state().activeStepIndex).toEqual(0);
    component.instance().onStepChanged(1);
    expect(component.state().activeStepIndex).toEqual(0);
  });

  it('checks initial values', () => {
    const component = shallow(testCreateVmWizard());
    expect(component.find(WizardPattern).props().nextStepDisabled).toBeTruthy();
    expect(component.find(WizardPattern).props().nextText).toBe('Next');
    expect(component.find(WizardPattern).props().steps).toHaveLength(ALL_TABS.length);
    expect(component.instance().getLastStepIndex()).toBe(ALL_TABS.length - 1);
  });

  it('changes next step disability', () => {
    const component = shallow(testCreateVmWizard());

    component.instance().onStepDataChanged(validBasicSettings, true);
    expect(component.state().stepData[0].value).toEqual(validBasicSettings);
    expect(component.state().stepData[0].valid).toBeTruthy();
    expect(component.find(WizardPattern).props().nextStepDisabled).toBeFalsy();

    // new required field will become visible
    component.instance().onStepDataChanged(
      {
        ...validBasicSettings,
        imageSourceType: {
          value: 'URL'
        }
      },
      false
    );
    expect(component.find(WizardPattern).props().nextStepDisabled).toBeTruthy();
  });

  it('creates vm', () => {
    createVM.mockReturnValueOnce(new Promise((resolve, reject) => resolve({ result: 'VM created' })));
    testWalkThrough();
  });

  it('fails creating the vm', () => {
    createVM.mockReturnValueOnce(new Promise((resolve, reject) => reject(new Error('VM not created'))));
    testWalkThrough();
  });
});
