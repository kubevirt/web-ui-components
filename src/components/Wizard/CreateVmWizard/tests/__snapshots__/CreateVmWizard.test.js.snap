// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CreateVmWizard /> changes next step disability 1`] = `
Object {
  "authKeys": Object {
    "isHidden": Object {
      "cloudInit": true,
      "useCloudInitCustomScript": false,
    },
    "value": null,
  },
  "cloudInit": Object {
    "isDisabled": Object {
      "VMWARE_PROVIDER_METADATA_ID": false,
    },
    "value": undefined,
  },
  "cloudInitCustomScript": Object {
    "isHidden": Object {
      "cloudInit": true,
      "useCloudInitCustomScript": true,
    },
    "value": null,
  },
  "containerImage": Object {
    "isDisabled": Object {
      "userTemplate": false,
    },
    "isHidden": Object {
      "provisionSourceType": false,
    },
    "isRequired": Object {
      "provisionSourceType": true,
    },
    "validation": null,
    "value": "pathtoimage",
  },
  "cpu": Object {
    "isDisabled": Object {
      "VMWARE_PROVIDER_METADATA_ID": false,
    },
    "isHidden": Object {
      "flavor": true,
    },
    "isRequired": Object {
      "flavor": false,
    },
    "value": undefined,
  },
  "description": Object {
    "isDisabled": Object {
      "VMWARE_PROVIDER_METADATA_ID": false,
    },
    "value": undefined,
  },
  "flavor": Object {
    "flavors": Array [
      "small",
      "Custom",
    ],
    "isDisabled": Object {
      "VMWARE_PROVIDER_METADATA_ID": false,
      "VM_SETTINGS_METADATA_ID": false,
    },
    "validation": undefined,
    "value": "small",
  },
  "hostname": Object {
    "isHidden": Object {
      "cloudInit": true,
      "useCloudInitCustomScript": false,
    },
    "value": null,
  },
  "imageURL": Object {
    "isDisabled": Object {
      "userTemplate": false,
    },
    "isHidden": Object {
      "provisionSourceType": true,
    },
    "isRequired": Object {
      "provisionSourceType": false,
    },
    "validation": null,
    "value": null,
  },
  "memory": Object {
    "isDisabled": Object {
      "VMWARE_PROVIDER_METADATA_ID": false,
    },
    "isHidden": Object {
      "flavor": true,
    },
    "isRequired": Object {
      "flavor": false,
    },
    "validation": null,
    "value": undefined,
  },
  "name": Object {
    "isDisabled": Object {
      "VMWARE_PROVIDER_METADATA_ID": false,
    },
    "validation": null,
    "value": "name",
  },
  "namespace": Object {
    "validation": undefined,
    "value": "namespace",
  },
  "operatingSystem": Object {
    "isDisabled": Object {
      "VMWARE_PROVIDER_METADATA_ID": false,
      "userTemplate": false,
    },
    "operatingSystems": Array [
      Object {
        "id": "fedora29",
        "name": "Fedora 29",
      },
      Object {
        "id": "fedora28",
        "name": "Fedora 28",
      },
      Object {
        "id": "fedora27",
        "name": "Fedora 27",
      },
      Object {
        "id": "fedora26",
        "name": "Fedora 26",
      },
      Object {
        "id": "fedora25",
        "name": "Fedora 25",
      },
      Object {
        "id": "fedora24",
        "name": "Fedora 24",
      },
      Object {
        "id": "fedora23",
        "name": "Fedora 23",
      },
      Object {
        "id": "rhel7.0",
        "name": "Red Hat Enterprise Linux 7.0",
      },
      Object {
        "id": "ubuntu18.04",
        "name": "Ubuntu 18.04 LTS",
      },
    ],
    "validation": undefined,
    "value": Object {
      "id": "fedora29",
      "name": "Fedora 29",
    },
  },
  "provider": Object {
    "isDisabled": Object {
      "provider": false,
    },
    "isHidden": Object {
      "provisionSourceType": true,
    },
    "isRequired": Object {
      "provisionSourceType": false,
    },
    "validation": null,
    "value": null,
  },
  "providersData": Object {
    "VMware": Object {
      "PROVIDER_VMWARE_V2V_LAST_ERROR": Object {
        "isDisabled": Object {
          "VMWARE_PROVIDER_METADATA_ID": false,
        },
        "isHidden": Object {
          "VMWARE_PROVIDER_METADATA_ID": true,
        },
      },
      "checkConnectionButton": Object {
        "isDisabled": Object {
          "VMWARE_PROVIDER_METADATA_ID": false,
          "vCenterInstance": true,
        },
        "isHidden": Object {
          "VMWARE_PROVIDER_METADATA_ID": true,
          "vCenterInstance": true,
        },
      },
      "checkConnectionButtonText": "Check",
      "rememberVMwareCredentials": Object {
        "isDisabled": Object {
          "VMWARE_PROVIDER_METADATA_ID": false,
        },
        "isHidden": Object {
          "VMWARE_PROVIDER_METADATA_ID": true,
          "vCenterInstance": true,
        },
      },
      "vCenterInstance": Object {
        "isDisabled": Object {
          "VMWARE_PROVIDER_METADATA_ID": false,
        },
        "isHidden": Object {
          "VMWARE_PROVIDER_METADATA_ID": true,
        },
        "isRequired": Object {
          "VMWARE_PROVIDER_METADATA_ID": false,
        },
      },
      "vmwareHostname": Object {
        "isDisabled": Object {
          "VMWARE_PROVIDER_METADATA_ID": false,
        },
        "isHidden": Object {
          "VMWARE_PROVIDER_METADATA_ID": true,
          "vCenterInstance": true,
        },
        "isRequired": Object {
          "vCenterInstance": false,
        },
      },
      "vmwarePassword": Object {
        "isDisabled": Object {
          "VMWARE_PROVIDER_METADATA_ID": false,
        },
        "isHidden": Object {
          "VMWARE_PROVIDER_METADATA_ID": true,
          "vCenterInstance": true,
        },
        "isRequired": Object {
          "vCenterInstance": false,
        },
      },
      "vmwareStatus": Object {
        "isHidden": Object {
          "VMWARE_PROVIDER_METADATA_ID": true,
        },
      },
      "vmwareUserName": Object {
        "isDisabled": Object {
          "VMWARE_PROVIDER_METADATA_ID": false,
        },
        "isHidden": Object {
          "VMWARE_PROVIDER_METADATA_ID": true,
          "vCenterInstance": true,
        },
        "isRequired": Object {
          "vCenterInstance": false,
        },
      },
      "vmwareVm": Object {
        "isDisabled": Object {
          "VMWARE_PROVIDER_METADATA_ID": true,
          "vmwareVm": true,
        },
        "isHidden": Object {
          "VMWARE_PROVIDER_METADATA_ID": true,
        },
        "isRequired": Object {
          "VMWARE_PROVIDER_METADATA_ID": false,
        },
      },
    },
  },
  "provisionSourceType": Object {
    "isDisabled": Object {
      "userTemplate": false,
    },
    "validation": undefined,
    "value": "Container",
  },
  "startVM": Object {
    "isHidden": Object {
      "provisionSourceType": false,
    },
    "value": undefined,
  },
  "useCloudInitCustomScript": Object {
    "isHidden": Object {
      "cloudInit": true,
    },
    "value": false,
  },
  "userTemplate": Object {
    "isHidden": Object {
      "provisionSourceType": false,
    },
    "value": undefined,
  },
  "workloadProfile": Object {
    "isDisabled": Object {
      "VMWARE_PROVIDER_METADATA_ID": false,
      "userTemplate": false,
    },
    "validation": undefined,
    "value": "generic",
    "workloadProfiles": Array [
      "generic",
    ],
  },
}
`;

exports[`<CreateVmWizard /> renders correctly 1`] = `
<WizardPattern
  activeStepIndex={0}
  backText="Back"
  backdrop="static"
  bodyHeader={null}
  cancelButtonDisabled={false}
  cancelText="Cancel"
  closeText="Close"
  dialogClassName="modal-lg wizard-pf kubevirt-wizard kubevirt-create-vm-wizard"
  loading={false}
  loadingMessage="Loading..."
  loadingTitle="Loading Wizard..."
  nextButtonRef={[Function]}
  nextStepDisabled={true}
  nextText="Next"
  onBack={[Function]}
  onExited={[Function]}
  onHide={[Function]}
  onNext={[Function]}
  onStepChanged={[Function]}
  previousStepDisabled={false}
  show={true}
  stepButtonsDisabled={false}
  steps={
    Array [
      Object {
        "key": "vmSettings",
        "onCloseWizard": [Function],
        "render": [Function],
        "title": "Basic Settings",
      },
      Object {
        "key": "network",
        "render": [Function],
        "title": "Networking",
      },
      Object {
        "key": "storage",
        "render": [Function],
        "title": "Storage",
      },
      Object {
        "key": "result",
        "render": [Function],
        "title": "Result",
      },
    ]
  }
  title="Create Virtual Machine"
/>
`;
